src/main/java/edu/ie3/datamodel/io/source/csv/CsvDataSource.java

```diff
@@ -360,61 +357,95 @@ protected List<Map<String, String>> csvRowFieldValueMapping(
+   * @param entityName name of the entity
+   * @param allRows of the csv file
+   * @param fieldsMap field name to fields
+   * @return a try object
    */
-  protected Try<Set<Map<String, String>>, SourceException> distinctRowsWithLog(
+  protected Try<Stream<Map<String, String>>, SourceException> checkUniqueness(
+      String entityName,
       Collection<Map<String, String>> allRows,
-      final Function<Map<String, String>, String> keyExtractor,
-      String entityDescriptor,
-      String keyDescriptor) {
-    Set<Map<String, String>> allRowsSet = new HashSet<>(allRows);
+      Map<String, Set<String>> fieldsMap) {
+    Set<Map<String, String>> rows = new HashSet<>(allRows);
+
     // check for duplicated rows that match exactly (full duplicates) -> sanity only, not crucial -
-    // case a)
-    if (allRows.size() != allRowsSet.size()) {
+    if (allRows.size() != rows.size()) {
       log.warn(
           "File with {} contains {} exact duplicated rows. File cleanup is recommended!",
-          entityDescriptor,
-          (allRows.size() - allRowsSet.size()));
+          entityName,
+          (allRows.size() - rows.size()));
+    }
+
+    List<SourceException> exceptions =
+        Try.getExceptions(
+            fieldsMap.entrySet().stream()
+                .map(e -> checkUniqueness(entityName, rows, e.getKey(), e.getValue())));
+
+    if (exceptions.isEmpty()) {
+      return Success.of(rows.parallelStream());
+    } else {
+      return Failure.of(
+          new SourceException(
+              "The following exception(s) occurred while checking the uniqueness of '"
+                  + entityName
+                  + "' entities: "
+                  + ExceptionUtils.getMessages(exceptions)));
     }
+  }
 
-    /* Check for rows with the same key based on the provided key extractor function */
-    Set<Map<String, String>> distinctIdSet =
-        allRowsSet.parallelStream()
-            .filter(ValidationUtils.distinctByKey(keyExtractor))
-            .collect(Collectors.toSet());
-    if (distinctIdSet.size() != allRowsSet.size()) {
-      allRowsSet.removeAll(distinctIdSet);
-      String affectedCoordinateIds =
-          allRowsSet.stream().map(keyExtractor).collect(Collectors.joining(",\n"));
-
-      // if this happens, we return a failure
+  /**
+   * Checking the uniqueness for a given entity.
+   *
+   * @param entityName name of the entity
+   * @param rows of the csv file
+   * @param fieldName name of the field
+   * @param fields set of strings used to create an extractor
+   * @return a try object
+   */
+  protected Try<Void, SourceException> checkUniqueness(
+      String entityName,
+      Collection<Map<String, String>> rows,
+      String fieldName,
+      Set<String> fields) {
+
+    Function<Map<String, String>, String> extractor =
+        fieldsToAttributes ->
+            fields.stream().map(fieldsToAttributes::get).collect(Collectors.joining(""));
```

Comment:
I think building a String out of all values is not better than calculating a hashcode - it'd really be best to just check unique fields only

Reply:
Let me work out a suggestion...

Reply:
This is used if we need the combintion of two fields (e.g. latitude/longitude).

Reply:
Ok, I introduced `List<Set<String>> getUniqueFields()` in `Factory` now (and also realized factories are nowhere near in sight where we need them here, but that's a different problem)

I also understand now how your comparison method works, and it's actually close to what I thought of, as well. What would be even easier (and faster, I think) is just creating sets of values that are supposed to be unique. So, when there are fields uuid and id that are supposed to be unique in combination, you just create a set of the given uuid and id for each row and compare uniqueness amongst all of them.

Reply:
So You mean like List.of(Set.of("uuid", "id")) instead of "uuidid"?

Reply:
Maybe passing through the List<Set<String>> uniqueFields down to `buildStreamWithFieldsToAttributesMap` is the way... I'm not quite sure yet

Reply:
> So you mean like List.of(Set.of("uuid", "id")) instead of "uuidid"?

Yes, exactly

Reply:
Where `uuid` and `id` are the respective values, of course. So it'd be something like `Set.of("79b6bcc0-56c7-4e9f-a4d9-e9c7e1482dbd", "entity_name")` for a single entity

Reply:
Or even better, a `Map<String, String>` from field name to value? Should work as well